SQL:Docs: add the PIVOT clause to SELECT section

bpintea opened the pull request:
This PR adds the documentation on the PIVOT clause.

bpintea committed:
SQL:Docs: add the PIVOT clause to SELECT section

The commit adds the documentation on the PIVOT clause.

elasticmachine commented:
Pinging @elastic/es-docs (>docs)

elasticmachine commented:
Pinging @elastic/es-search (:Search/SQL)

costin reviewed (approved):
LGTM. The examples look acceptable so we don't have to add yet another one.
In a follow-up maybe it would help to add more explanations on the result, pointing out how the aggregation happens on and where the pivot comes in.
If you are familiar with PIVOT things are obvious but wrapping your mind at it, at the beginning takes some time and an explanation of the example should help.

matriv commented on docs/reference/sql/language/syntax/commands/select.asciidoc:
```diff
@@ -391,3 +392,54 @@ To return
+include-tagged::{sql-specs}/pivot.csv-spec[sumWithoutSubquery]
+----
+
+The query could logically be broken down in the following steps:
```
Maybe the `query execution` ?

bpintea replied:
Thanks, added.

matriv commented on docs/reference/sql/language/syntax/commands/select.asciidoc:
```diff
@@ -391,3 +392,54 @@ To return
+. the resulting values are filtered through the set provided in the `IN`-clause;
+. the now filtered column is pivoted to form the headers of the two additional columns appended to the result: `1` and `2`;
+. a <<sql-syntax-group-by, GROUP BY>> on all columns of the source table `test_emp`, except `salary` (part of the aggregation subclause) and `languages` (part of the `FOR`-clause);
+. the values in these appended columns are the `SUM` aggregations of `salary`, group by the respective language.
```
```suggestion
. the values in these appended columns are the `SUM` aggregations of `salary`, grouped by the respective language.
```

bpintea replied:
fixed.

matriv reviewed (commented):
Looks good!
Left a couple of suggestions, and a question: Should we add something in the limitations page regarding the IN (no subquery supported) and that only pivoting by 1 column is currently allowed?

bpintea committed:
address review suggestions

Add small changes, as suggested in the PR review.

bpintea reviewed (commented):
> Left a couple of suggestions, and a question: Should we add something in the limitations page regarding the IN (no subquery supported) and that only pivoting by 1 column is currently allowed?

The subquery for `IN` limitation is ~more~ **less** common, I think. The single pivoting column one could be added, indeed. Unless we're going to lift that anytime soon? @costin, wdyt?

astefan reviewed (approved):
LGTM, left some very minor comments.

bpintea committed:
address review suggestions

break sentence with `:` before enumeration.

bpintea commented on docs/reference/sql/language/syntax/commands/select.asciidoc:
```diff
@@ -391,3 +392,54 @@ To return
+
+The `PIVOT` clause performs a cross tabulation on the results of the query: it aggregates the results and rotates rows into columns. The rotation is done by turning unique values from one column in the expression - the pivoting column - into multiple columns in the output. The column values are aggregations on the remaining columns specified in the expression.
+
+The clause can be broken down in three parts, the aggregation, the `FOR`- and the `IN`-subclause.
```
thanks, fixed.

bpintea commented on docs/reference/sql/language/syntax/commands/select.asciidoc:
```diff
@@ -391,3 +392,54 @@ To return
+
+The `FOR`-subclause specifies the pivoting column: the distinct values of this column will become the candidate set of values to be rotated.
+
+The `IN`-subclause defines a filter: the intersection between the set provided here and the candidate set from the `FOR`-subclause will be rotated to become the headers of the columns appended to the end result. The filter can not be a subquery, one must provide here literal values, obtained in advance.
```
I'd keep the phrasing, as they are in fact literals and while there is a meanings intersection with "constant", there is no identity.

bpintea committed:
add PIVOT limitations to the corresponding page

- address review suggestions;

- also, s/retrieveable/retrievable in 'Retrieving from docvalue_fields'
paragraph.

matriv commented on docs/reference/sql/limitations.asciidoc:
```diff
@@ -176,6 +176,25 @@ from `docvalue_fields`.
+SELECT * FROM test_emp PIVOT (SUM(salary) FOR languages IN (1, 2))
+-------------------------------------------------------------
+
+the `languages` of interest must be listed explicitly: `IN (1, 2)`.
```
Maybe also provide a real subquery example to mention that it doesn't work.

bpintea replied:
Thanks. I did.

matriv commented on docs/reference/sql/limitations.asciidoc:
```diff
@@ -176,6 +176,25 @@ from `docvalue_fields`.
+[[aggs-in-pivot]]
+=== Aggregations in the <<sql-syntax-pivot, `PIVOT`>> clause
+
+The aggregation expression in <<sql-syntax-pivot, `PIVOT`>> will currently accept only one aggregation. It is thus not possible to obtain multiple aggregations for any one pivoted columns.
```
`...to obtain multiple aggregations for any one pivoted columns.` something is missing here?
Also I'm not native english speaker but how about: `Thus, it is not possible...`

bpintea replied:
> `...to obtain multiple aggregations for any one pivoted columns.` something is missing here?

One extra trailing `s` left after rephrasing. Removed it.

> `Thus, it is not possible...`

That'd also be possible, but a matter of taste. I'd thus leave it as is. :-)

matriv reviewed (approved):
LGTM. Left a couple of comments.

bpintea committed:
add counter-example for PIVOT subquery

- addressing review suggestions.

matriv commented on docs/reference/sql/limitations.asciidoc:
```diff
@@ -197,4 +197,9 @@ The values that the <<sql-syntax-pivot, `PIVOT`>> query could pivot must be prov
+
+[source, sql]
+-------------------------------------------------------------
+SELECT * FROM test_emp PIVOT (SUM(salary) FOR languages IN (SELECT languages FROM test_emp WHERE languages <=2 GROUP BY languages))
```
I think it would be better to have something like: ```... IN (SELECT languages FROM test_emp WHERE salary > 10000)```

bpintea replied:
I wanted to have a subquery that would produce the the same result set as the working query.
But what's the thought behind your example?

matriv replied:
I get your point, my thinking is that this feels like artificial and not "real life" query.

bpintea replied:
I see. I would hope that the "working" example meets the "realistic" requirements.
The proposed alternative for the "not working" is less convoluted, true, but IMO it would required support for "unique" / "distinct" to bring it closer to the results of the "working" example and the implicit grouping that actually PIVOT does, before applying the IN-filtering.

matriv commented on docs/reference/sql/limitations.asciidoc:
```diff
@@ -184,7 +184,7 @@ setting needs to be adjusted or fewer columns retrievable from `docvalue_fields`
 === Aggregations in the <<sql-syntax-pivot, `PIVOT`>> clause
 
-The aggregation expression in <<sql-syntax-pivot, `PIVOT`>> will currently accept only one aggregation. It is thus not possible to obtain multiple aggregations for any one pivoted columns.
+The aggregation expression in <<sql-syntax-pivot, `PIVOT`>> will currently accept only one aggregation. It is thus not possible to obtain multiple aggregations for any one pivoted column.
```
Have the feeling that `one` is superfluous but might be wrong.

bpintea replied:
I believe that the `for any one <item>` should be a valid phrasing.

bpintea commented:
Ports:
7.x: 8c2ab8bb721
7.5: 527e21552b8
